dbms/programs/client/Client.cpp

```diff
@@ -89,6 +89,44 @@
 #define DISABLE_LINE_WRAPPING "\033[?7l"
 #define ENABLE_LINE_WRAPPING "\033[?7h"
 
+#if USE_READLINE && RL_VERSION_MAJOR >= 7
+
+#define BRACK_PASTE_PREF "\033[200~"
+#define BRACK_PASTE_SUFF "\033[201~"
+
+#define BRACK_PASTE_LAST '~'
+#define BRACK_PASTE_SLEN 6
+
+
+/// Since we've already joined multiple lines when putting queries into history, make sure pasting
+///  always have lines joined to provide a consistent user experience. Otherwise some terminals
+///  might generate ^J for the enter character while others gives multiple lines without prompt.
+///  This handler also bypasses some unused macro/event checkings.
+extern "C" int clickhouse_rl_bracketed_paste_begin(int, int)
+{
+    std::string buf;
+    buf.reserve(128);
+
+    RL_SETSTATE(RL_STATE_MOREINPUT);
+    char c;
+    while ((c = rl_read_key()) >= 0)
+    {
+        if (c == '\r' || c == '\n')
+            c = ' ';
+        buf.push_back(c);
+        if (buf.size() >= BRACK_PASTE_SLEN && c == BRACK_PASTE_LAST && buf.substr(buf.size() - BRACK_PASTE_SLEN) == BRACK_PASTE_SUFF)
+        {
+            buf.resize(buf.size() - BRACK_PASTE_SLEN);
+            break;
+        }
+    }
+    RL_UNSETSTATE(RL_STATE_MOREINPUT);
```

Comment:
Should we use SCOPE_EXIT to protect from possible exceptions from buf.resize and buf.substr? (std::bad_alloc)